### PR TITLE
feat(list): add list_filter expression

### LINKED ISSUE #4486: `expr.list.map(...)` and `expr.list.filter(...)`
### Is your feature request related to a problem?

oftentimes when working with nested data, i want to perform common functional operations on those list columns such as (map, filter, ...)

### Describe the solution you'd like

So i think there's 2 ways we could approach this. 

### 1. via lambdas
```py
df = daft.from_pydict({"text": [["hello", "world"], ["goodbye", "world"]]}
df.select(col("text").list.map(lambda x: x.str.upper()))
```
While I think the interface for this is a bit nicer, it also makes it a lot harder to expose this to sql, and makes optimization much harder

### 2. via a new `.element` type
```py
df = daft.from_pydict({"text": [["hello", "world"], ["goodbye", "world"]]}
df.select(col("text").list.map(daft.element().str.upper())) # daft.element() is a special kind of `expr` that is just a placeholder only available within nested contexts. 

```
This is inspired by how polars allows you to do list operations without needing to resort to python lambdas. This also gives us much more control within the optimizer as everything is still expressed as an expression, and we don't need to call into python. 

this second approach also allows us to expose similar functionality through sql lambda functions. 

```sql
select list_map(text, x -> upper(x)) from df;
select list_filter(text, x -> x <> 'world') from df;

```

### Describe alternatives you've considered

udf's are currently the only way you can map or filter elements inside of a list.

### Additional Context

_No response_

### Would you like to implement a fix?

Yes

### PR DESCRIPTION
**Pull request by aaron-ang**
## Changes Made

Adds `expr.list_filter(predicate)` and `daft.functions.list_filter`, which filter list elements by a Boolean predicate over `daft.element()`. SQL form `list_filter(col, element() != 'x')` works automatically through the function registry.

Behavior:
- Null list rows pass through as null.
- Empty list rows pass through as empty.
- Null predicate elements drop the element (matches SQL `WHERE`).

One incidental fix: `resolve_list_evals` now explodes with `ignore_empty_and_null=true` so the inner sub-expression aligns 1:1 with `flat_child`. Without this, an empty or null list row injected a placeholder that shifted positions and misaligned the mask which is required for `list_filter`, and also fixes a latent alignment bug in `list_map` when an empty/null row appeared before any non-empty rows.

`split_udfs`'s existing "do not split UDFs inside `list_map`" bail-out was generalized to cover `list_filter` as well (same reason. UDFs inside list-eval operate on the element scope and can't be lifted to the row scope without a plan-topology rewrite).

## Related Issues

Closes #4486

### CODE REVIEW

**Comment by aaron-ang** [comment]
@greptile-apps re-review

File: src/daft-functions-list/src/list_filter.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        _ctx: &daft_dsl::functions::scalar::EvalContext,
+    ) -> DaftResult<Series> {
+        // Predicate children are evaluated first, so `mask` is already a flat boolean
+        // series aligned 1:1 with `flat_child`. Nulls in the predicate drop the element.
+        let ListFilterArgs {
+            input: list_arr,
+            expr: mask,
+        } = inputs.try_into()?;
+
+        let list_arr = list_arr.list()?;
```
[the remark this answers is not in the document]
Comment (aaron-ang) [reply src/daft-functions-list/src/list_filter.rs:32]: removed `|| input_field.dtype.is_fixed_size_list()` check.

File: tests/functions/test_list_filter.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            "structs": [
+                [{"a": 1, "b": "x"}, {"a": 5, "b": "y"}],
+                [{"a": 0, "b": "z"}],
+                [],
+            ]
+        }
+    )
+    res = df.select(daft.col("structs").list_filter(daft.element().get("a") > 0)).to_pydict()["structs"]
+    assert res == [[{"a": 1, "b": "x"}, {"a": 5, "b": "y"}], [], []]
```
[the remark this answers is not in the document]
Comment (aaron-ang) [reply tests/functions/test_list_filter.py:89]: added `test_list_filter_null_element`

**Comment by rohitkulshreshtha** [comment]
Thanks @aaron-ang — clean addition, and I appreciated the care in fixing the latent `list_map` alignment bug as part of this PR (and adding regression tests for both empty-first / null-middle positions — right instinct). The explode-then-mask pipeline with `ignore_empty_and_null=true` is the correct approach, and the `split_udfs` generalization mirrors the existing ListMap pattern cleanly.

Two test gaps worth filling, non-blocking:

1. **Non-Boolean predicate → TypeError.** `get_return_field` enforces the predicate dtype must be `Boolean`, but there's no test that passes a non-Boolean predicate and asserts the error. Would guard the type-check from regressing.

2. **Nested-list (List of Lists) test.** Does `daft.element()` bind correctly when the inner type is itself a list? A single parametrize or dedicated test would cover this cleanly.

LGTM on merge once those (or a follow-up commit covering them) land — happy to re-look if you drop a new commit.

**Comment by aaron-ang** [comment]
> Thanks @aaron-ang — clean addition, and I appreciated the care in fixing the latent `list_map` alignment bug as part of this PR (and adding regression tests for both empty-first / null-middle positions — right instinct). The explode-then-mask pipeline with `ignore_empty_and_null=true` is the correct approach, and the `split_udfs` generalization mirrors the existing ListMap pattern cleanly.
> 
> 
> 
> Two test gaps worth filling, non-blocking:
> 
> 
> 
> 1. **Non-Boolean predicate → TypeError.** `get_return_field` enforces the predicate dtype must be `Boolean`, but there's no test that passes a non-Boolean predicate and asserts the error. Would guard the type-check from regressing.
> 
> 
> 
> 2. **Nested-list (List of Lists) test.** Does `daft.element()` bind correctly when the inner type is itself a list? A single parametrize or dedicated test would cover this cleanly.
> 
> 
> 
> LGTM on merge once those (or a follow-up commit covering them) land — happy to re-look if you drop a new commit.
> 
> 

hi @rohitkulshreshtha, thanks for the comments. i will add a commit to address them.

**Comment by rohitkulshreshtha** [comment]
Thanks @aaron-ang — both gaps are covered cleanly: the non-Boolean predicate test exercises the dtype guard end-to-end, and the nested-list case confirms `daft.element()` binds correctly through a `List<List<...>>`. LGTM. Will merge through the `rust-dependency-budget` fast-fail.

**Review by rohitkulshreshtha — APPROVED** [review]
